File: python/medic/plugins/Tester/laminaFace.py
from medic.core import testerBase
from maya import OpenMaya


class LaminaFace(testerBase.TesterBase):
    Name = "LaminaFace"

    def __init__(self):
        super(LaminaFace, self).__init__()

    def Match(self, node):
        return node.object().hasFn(OpenMaya.MFn.kMesh)

    def Test(self, node):
        it = None
        try:
            it = OpenMaya.MItMeshPolygon(node.object())
        except:
            return (False, None)

        result = False

        comp = OpenMaya.MFnSingleIndexedComponent()
        comp_obj = comp.create(OpenMaya.MFn.kMeshPolygonComponent)

        while (not it.isDone()):
            if it.isLamina():
                comp.addElement(it.index())
                result = True

            it.next()

        return (result, comp_obj if result else None)


Tester = LaminaFace

File: plugins/Tester/laminaFace.cpp
#include "medic/pluginapi.h"
#include <maya/MFn.h>
#include <maya/MStatus.h>
#include <maya/MObject.h>
#include <maya/MItMeshPolygon.h>
#include <maya/MFnSingleIndexedComponent.h>


using namespace MEDIC;


class LaminaFace : public MdTester
{
    std::string Name();
    std::string Description();
    bool Match(MdNode *node);
    MdReport *test(MdNode *node);
};


std::string LaminaFace::Name()
{
    return "LaminaFace";
}

std::string LaminaFace::Description()
{
    return "Mesh has lamina face(s)";
}

bool LaminaFace::Match(MdNode *node)
{
    return node->object().hasFn(MFn::kMesh);
}

MdReport *LaminaFace::test(MdNode *node)
{
    MStatus stat;
    MItMeshPolygon it(node->object(), &stat);

    if (stat == MS::kFailure)
    {
        return 0;
    }

    bool result = false;
    MFnSingleIndexedComponent comp;
    MObject comp_obj = comp.create(MFn::kMeshPolygonComponent);

    while (!it.isDone())
    {
        if (it.isLamina())
        {
            result = true;
            comp.addElement(it.index());
        }

        it.next();
    }

    if (!result)
    {
        return 0;
    }

    return new MdReport(node, comp_obj);
}


MEDIC_PLUGIN_API MdTester *Create()
{
    return new LaminaFace();
}